### PR TITLE
feat/amazon_support

### PR DESCRIPTION
**Pull request by alhiwatan**
Add better support for Amazon bidding.

Ported from the React Native PR [#291](https://github.com/AppLovin/AppLovin-MAX-React-Native/pull/291).

Tested both iOS and Android.

Calling methods are the same as the RN ones.

**iOS**
- Banners
```
        Class AppLovinMAXClass = NSClassFromString(@"AppLovinMAX");
        NSObject *AppLovinMAX = [AppLovinMAXClass performSelector: @selector(shared)];
        [AppLovinMAX performSelector: @selector(setAmazonResult:forBannerAdUnitIdentifier:)
                                            withObject: result
                                            withObject: @"YOUR_AD_UNIT_ID"];
```
- MRECs
```
        Class AppLovinMAXClass = NSClassFromString(@"AppLovinMAX");
        NSObject *AppLovinMAX = [AppLovinMAXClass performSelector: @selector(shared)];
        [AppLovinMAX performSelector: @selector(setAmazonResult:forMRecAdUnitIdentifier:)
                                            withObject: result
                                            withObject: @"YOUR_AD_UNIT_ID"];
```
- Interstitials
```
        Class AppLovinMAXClass = NSClassFromString(@"AppLovinMAX");
        NSObject *AppLovinMAX = [AppLovinMAXClass performSelector: @selector(shared)];
        [AppLovinMAX performSelector: @selector(setAmazonResult:forInterstitialAdUnitIdentifier:)
                                            withObject: result
                                            withObject: @"YOUR_AD_UNIT_ID"];
```
**Android**
- Banners
```
try
{
    Class<?> clazz = Class.forName( "com.applovin.applovin_max.AppLovinMAX" );
    Method getInstanceMethod = clazz.getMethod( "getInstance" );
    Object appLovinMAXModule = getInstanceMethod.invoke( null );
    Method setAmazonBannerResultMethod = clazz.getMethod( "setAmazonBannerResult", Object.class, String.class );
    setAmazonBannerResultMethod.invoke( appLovinMAXModule, result, "YOUR_AD_UNIT_ID" );
}
catch ( ClassNotFoundException e ) { }
catch ( NoSuchMethodException e ) { }
catch ( IllegalAccessException e ) { }
catch ( InvocationTargetException e ) { }
```
- MRECs
```
try
{
    Class<?> clazz = Class.forName( "com.applovin.applovin_max.AppLovinMAX" );
    Method getInstanceMethod = clazz.getMethod( "getInstance" );
    Object appLovinMAXModule = getInstanceMethod.invoke( null );
    Method setAmazonMRecResultMethod = clazz.getMethod( "setAmazonMRecResult", Object.class, String.class );
    setAmazonMRecResultMethod.invoke( appLovinMAXModule, result, "YOUR_AD_UNIT_ID" );
}
catch ( ClassNotFoundException e ) { }
catch ( NoSuchMethodException e ) { }
catch ( IllegalAccessException e ) { }
catch ( InvocationTargetException e ) { }
```
- Interstitials
```
try
{
    Class<?> clazz = Class.forName( "com.applovin.applovin_max.AppLovinMAX" );
    Method getInstanceMethod = clazz.getMethod( "getInstance" );
    Object appLovinMAXModule = getInstanceMethod.invoke( null );
    Method setAmazonInterstitialResultMethod = clazz.getMethod( "setAmazonInterstitialResult", Object.class, String.class );
    setAmazonInterstitialResultMethod.invoke( appLovinMAXModule, result, "YOUR_AD_UNIT_ID" );
}
catch ( ClassNotFoundException e ) { }
catch ( NoSuchMethodException e ) { }
catch ( IllegalAccessException e ) { }
catch ( InvocationTargetException e ) { }
```

### CODE REVIEW

File: applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java [hunk 1]
```diff
@@ -1678,6 +1678,63 @@ private Map<String, Object> createNetworkResponseInfo(final MaxNetworkResponseIn
...
+            MaxInterstitialAd interstitial = retrieveInterstitial( adUnitId );
+            if ( interstitial == null )
+            {
+                e( "Unable to set Amazon result - unable to retrieve interstitial" );
+                return;
+            }
+
+            interstitial.setLocalExtraParameter( key, result );
+        }
+        else
```
Comment (santoshbagadi) [inline applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java:1725]:
```suggestion
        else // MaxAdFormat.BANNER or MaxAdFormat.MREC
```

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 2]
```diff
@@ -1612,6 +1612,61 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
...
+    }
+    
+    NSString *key = [self localExtraParameterKeyForAmazonResult: result];
+    
+    if ( adFormat == MAAdFormat.interstitial )
+    {
+        MAInterstitialAd *interstitial = [self retrieveInterstitialForAdUnitIdentifier: adUnitIdentifier];
+        [interstitial setLocalExtraParameterForKey: key value: result];
+    }
+    else
```
Comment (santoshbagadi) [inline applovin_max/ios/Classes/AppLovinMAX.m:1655]:
```suggestion
    else // MAAdFormat.banner or MAAdFormat.mrec
```

File: applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java [hunk 3]
```diff
@@ -1678,6 +1678,63 @@ private Map<String, Object> createNetworkResponseInfo(final MaxNetworkResponseIn
...
+        {
+            MaxAdView adView = retrieveAdView( adUnitId, adFormat );
+            adView.setLocalExtraParameter( key, result );
```
Comment (santoshbagadi) [inline applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAX.java:1728]: We should `null` check `adView` here and log an error similar to inters.

File: applovin_max/ios/Classes/AppLovinMAX.m [hunk 4]
```diff
@@ -1612,6 +1612,61 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
...
+    {
+        MAAdView *adView = [self retrieveAdViewForAdUnitIdentifier: adUnitIdentifier adFormat: adFormat];
+        [adView setLocalExtraParameterForKey: key value: result];
```
Comment (santoshbagadi) [inline applovin_max/ios/Classes/AppLovinMAX.m:1658]: Same here, should `nil` check `adView`.

**Review by thomasmso — APPROVED** [review]
@alhiwatan - Any updates you make from @santoshbagadi's suggestions, could you also port to RN?

**Comment by alhiwatan** [comment]
Updated both RN and Flutter.

**Review by santoshbagadi — APPROVED** [review]

**Pull request merged**